src/transforms/geoip.rs

```diff
@@ -0,0 +1,274 @@
+use super::Transform;
+
+use crate::{
+    event::{Event, ValueKind},
+    topology::config::{DataType, TransformConfig},
+};
+use serde::{Deserialize, Serialize};
+use string_cache::DefaultAtom as Atom;
+
+use std::str::FromStr;
+use tracing::field;
+
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct GeoipConfig {
+    pub source: Atom,
+    pub database: String,
+    #[serde(default = "default_geoip_target_field")]
+    pub target: String,
+}
+
+pub struct Geoip {
+    pub dbreader: maxminddb::Reader<Vec<u8>>,
+    pub source: Atom,
+    pub target: String,
+}
+
+fn default_geoip_target_field() -> String {
+    "geoip".to_string()
+}
+
+#[typetag::serde(name = "geoip")]
+impl TransformConfig for GeoipConfig {
+    fn build(&self) -> Result<Box<dyn Transform>, crate::Error> {
+        let reader = maxminddb::Reader::open_readfile(self.database.clone())?;
+        Ok(Box::new(Geoip::new(
+            reader,
+            self.source.clone(),
+            self.target.clone(),
+        )))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "geoip"
+    }
+}
+
+impl Geoip {
+    pub fn new(dbreader: maxminddb::Reader<Vec<u8>>, source: Atom, target: String) -> Self {
+        Geoip {
+            dbreader: dbreader,
+            source: source,
+            target: target,
+        }
+    }
+}
+
+impl Transform for Geoip {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        let target_field = self.target.clone();
+        let ipaddress = event
+            .as_log()
+            .get(&self.source)
+            .map(|s| s.to_string_lossy());
+        if let Some(ipaddress) = &ipaddress {
+            if let Ok(ip) = FromStr::from_str(ipaddress) {
+                if let Ok(data) = self.dbreader.lookup::<maxminddb::geoip2::City>(ip) {
+                    let city = data.city;
+                    if let Some(city) = city {
+                        let city_names = city.names;
+                        if let Some(city_names) = city_names {
+                            let city_name_en = city_names.get("en");
+                            if let Some(city_name_en) = city_name_en {
```

Comment:
Could we shorten this to something like:

```rust
if let (city_name_en) = data.city.and_then(|c| c.names).and_then(|names| names.get("en")) {
```

Reply:
Thanks @lukesteensen - I made an update. I couldn't successfully do away with all  the `if let` checks.  When I used the combinator expression for the `names.get()` bit as well, I am getting: 
```error[E0515]: cannot return value referencing function parameter `names`
  --> src/transforms/geoip.rs:79:43
   |
79 |                         .and_then(|names| names.get("en"))
   |                                           -----^^^^^^^^^^
   |                                           |
   |                                           returns a value referencing data owned by the current function
   |                                           `names` is borrowed here
```

Perhaps, you can suggest what I am doing wrong?

Reply:
That's my mistake, the example I gave was invalid. The rest looks great, thank you!